### PR TITLE
Put some space for non-assets requests in development mode

### LINKED ISSUE #23428: Better spacing with the default logger on Puma
The default logger needs to send a CR after the request is over. Right now it's all smushed together like:

```
Processing by PostsController#show as HTML
  Parameters: {"id"=>"1"}
  Post Load (0.1ms)  SELECT  "posts".* FROM "posts" WHERE "posts"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  Rendered posts/show.html.erb within layouts/application (0.6ms)
Completed 200 OK in 22ms (Views: 18.2ms | ActiveRecord: 0.1ms)
Started GET "/posts" for ::1 at 2016-02-02 16:06:30 +0100
Processing by PostsController#index as HTML
  Post Load (0.2ms)  SELECT "posts".* FROM "posts"
  Rendered posts/index.html.erb within layouts/application (1.9ms)
Completed 200 OK in 26ms (Views: 24.6ms | ActiveRecord: 0.2ms)
```

Should be:

```
Processing by PostsController#show as HTML
  Parameters: {"id"=>"1"}
  Post Load (0.1ms)  SELECT  "posts".* FROM "posts" WHERE "posts"."id" = ? LIMIT ?  [["id", 1], ["LIMIT", 1]]
  Rendered posts/show.html.erb within layouts/application (0.6ms)
Completed 200 OK in 22ms (Views: 18.2ms | ActiveRecord: 0.1ms)

Started GET "/posts" for ::1 at 2016-02-02 16:06:30 +0100
Processing by PostsController#index as HTML
  Post Load (0.2ms)  SELECT "posts".* FROM "posts"
  Rendered posts/index.html.erb within layouts/application (1.9ms)
Completed 200 OK in 26ms (Views: 24.6ms | ActiveRecord: 0.2ms)
```

### PR DESCRIPTION
**Pull request by prathamesh-sonpatki**
- Fixes #23428.

r? @dhh 

Here is the sample output on an app created from rails master with this change - https://gist.github.com/prathamesh-sonpatki/ca9dbc32bcbae9066562

~~EDIT: Updated gist with Kasper's suggestion of printing a line before starting assets request - https://gist.github.com/prathamesh-sonpatki/c9075f78ac6f1f372c7c~~

### CODE REVIEW

**Comment by kaspth** [comment]
Shouldn't one space be enough? Having two lines between requests feels airy to me :smile:

Can we make it so that there's a line just above the first asset request? Basically add a newline after: https://gist.github.com/prathamesh-sonpatki/ca9dbc32bcbae9066562#file-gistfile1-txt-L31

**Comment by prathamesh-sonpatki** [comment]
@kaspth like this - https://gist.github.com/prathamesh-sonpatki/c9075f78ac6f1f372c7c ?

**Comment by kaspth** [comment]
Almost! These should just be one line, IMO: https://gist.github.com/prathamesh-sonpatki/c9075f78ac6f1f372c7c#file-gistfile1-txt-L16-L17 😁

File: railties/lib/rails/rack/logger.rb [hunk 1]
```diff
@@ -30,6 +30,12 @@ def call(env)
     protected
 
       def call_app(request, env)
+        # Put some space between for non-assets requests in development logs.
+        if development? && request.path !~ /^\/assets/
+          logger.debug ''
+          logger.debug ''
```
Comment (nonmadden) [inline railties/lib/rails/rack/logger.rb]: why? `logger.debug ''`

Comment (kaspth) [inline railties/lib/rails/rack/logger.rb]: `# Put some space between for non-assets requests in development logs.` 😁

**Comment by prathamesh-sonpatki** [comment]
@kaspth For that we will need to know previous request path and based on that make a decision whether to show the line or not. I guess it will be more complicated.

**Comment by kaspth** [comment]
@prathamesh-sonpatki do we? Isn't outputting a space after every non asset request enough to get the result?

**Comment by prathamesh-sonpatki** [comment]
@kaspth That will work for non-asset version and asset version sequence but will not work for asset-request and non-asset request sequence.

For eg.

```
Started GET "/assets/application.self-b89234cf2659d7fedea75bca0b8d231ad7dfc2f3f57fcbaf5f44ed9dc384137b.js?body=1" for 127.0.0.1 at 2016-02-03 14:00:33 +0530
Started GET "/posts/new" for 127.0.0.1 at 2016-02-03 14:00:41 +0530
Processing by PostsController#new as HTML
  Rendered posts/_form.html.erb (11.2ms)
  Rendered posts/new.html.erb within layouts/application (13.5ms)
Completed 200 OK in 50ms (Views: 42.4ms | ActiveRecord: 3.2ms) (Views: 42.4ms | ActiveRecord: 3.2ms)
```

**Comment by prathamesh-sonpatki** [comment]
> I think definitely we should keep Puma to multi threads but let's only send the line clears together with the line for completing a Rails request (the line reporting status code and timing). Don't want to send it together with assets etc.

I also tried this suggestion from DHH in https://github.com/rails/rails/issues/23428. With this approach we will get same result as in https://gist.github.com/prathamesh-sonpatki/ca9dbc32bcbae9066562. There will be no extra line before assets request as present here - https://gist.github.com/prathamesh-sonpatki/c9075f78ac6f1f372c7c

File: railties/lib/rails/rack/logger.rb [hunk 2]
```diff
@@ -79,6 +82,12 @@ def development?
       def logger
         Rails.logger
       end
+
+      def show_extra_line_in_development_for_non_assets_requests(path)
+        if development? && path !~ /^\/assets/
```
Comment (vipulnsward) [inline railties/lib/rails/rack/logger.rb]: Does this work safely to handle public assets, assets from engines, vendored assets, etc?

btw, this won't work with http://guides.rubyonrails.org/asset_pipeline.html#changing-the-assets-path

Comment (prathamesh-sonpatki) [inline railties/lib/rails/rack/logger.rb]: Nice catch! It works with vendored assets but won't work with changing assets path.

**Comment by prathamesh-sonpatki** [comment]
I used DHH's original suggestion, it does not print a new line after all assets requests end but covers all other cases.

File: actionpack/lib/action_controller/log_subscriber.rb [hunk 3]
```diff
@@ -26,6 +26,9 @@ def process_action(event)
         end
         message = "Completed #{status} #{Rack::Utils::HTTP_STATUS_CODES[status]} in #{event.duration.round}ms"
         message << " (#{additions.join(" | ".freeze)})" unless additions.blank?
+        message << " (#{additions.join(" | ".freeze)})" unless additions.blank?
```
Comment (matthewd) [inline actionpack/lib/action_controller/log_subscriber.rb]: Don't think we need this :wink:

Comment (prathamesh-sonpatki) [inline actionpack/lib/action_controller/log_subscriber.rb]: Fixed :smile:

**Comment by kaspth** [comment]
:shipit:

**Comment by schneems** [comment]
:shipit:

File: actionpack/lib/action_controller/log_subscriber.rb [hunk 4]
```diff
@@ -26,6 +26,8 @@ def process_action(event)
         end
         message = "Completed #{status} #{Rack::Utils::HTTP_STATUS_CODES[status]} in #{event.duration.round}ms"
         message << " (#{additions.join(" | ".freeze)})" unless additions.blank?
+        message << "\n\n" if Rails.env.development?
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/log_subscriber.rb]: This is coupling Action Pack with Railties we should check if `Rails.env` is defined before calling it.

Comment (kaspth) [inline actionpack/lib/action_controller/log_subscriber.rb]: @rafaelfranca fixed @ https://github.com/rails/rails/commit/4e4bcae0841b81438ce3b6c627eebc7536566bf5

Comment (rafaelfranca) [inline actionpack/lib/action_controller/log_subscriber.rb]: Thanks!

**Pull request merged**